src/interpolate.rs

```diff
@@ -0,0 +1,210 @@
+use {Duplex, Frame, Sample};
+
+/// An iterator that converts the rate at which frames are yielded from some given frame
+/// Interpolator into a new type.
+///
+/// Other names for `sample::rate::Converter` might include:
+///
+/// - Sample rate converter
+/// - {Up/Down}sampler
+/// - Sample interpolater
+/// - Sample decimator
+///
+#[derive(Clone)]
+pub struct Converter<T: Iterator, I: Interpolator>
+    where <T as Iterator>::Item: Frame
+{
+    source: T,
+    interpolator: I,
+    interpolation_value: f64,
+    source_to_target_ratio: f64
+}
+
+/// Interpolator that just rounds off any values to the previous value from the source
+pub struct Floor<F>
+{
+    left: F
+}
+
+/// Interpolator that interpolates linearly between the previous value and the next value
+pub struct Linear<F>
+{
+    left: F,
+    right: Option<F>
+}
+
+/// Trait for all things that can interpolate between two values. Implementations should keep track
+/// of the necessary data both before and after the current frame.
+pub trait Interpolator
+{
+    type Frame: Frame;
+
+    /// Given a distance between [0. and 1.) to the following sample, return the interpolated value
+    fn interpolate(&self, x: f64) -> Self::Frame;
```

Comment:
It'd be nice if we could eventually work out some API that allowed for interpolating via integer ratios rather than `f64` to avoid the casts back and forth for signals of integer types like `i32` or `i16`. It's a bit of a pity that we require the `Duplex<f64>` bound on each of the `impl`s below as there are almost certainly more efficient methods of interpolation possible for each of the other Sample types. I'm not yet sure how we'd generalise this though.

I think the current `f64`-dependent impl results from  `Converter` storing the `source_to_target_ratio` as an `f64`, whereas it could just as easily store both the source and target as some type other than `f64` that is more efficient for interpolation of some other sample type. Will have to give this some more thought, though I think I'm fine with this as it is for now.